### PR TITLE
ADBDEV-4910-71: Move memcpy() under a check

### PR DESCRIPTION
**Pull request by bandetto**
Move memcpy() under a check

*p_grpColIdx was passed to memcpy() as a source before it was checked if it's
NULL. There is a NULL check right after.

In that condition, *p_grpColIdx is assumed to be non-NULL if *p_numGroupCols
is > 0.

Move memcpy() calls to be called only if *p_grpColIdx > 0 and add asserts for
*p_grpColIdx and *p_grpOperators.

### CODE REVIEW

File: src/backend/optimizer/plan/plangroupext.c [hunk 1]
```diff
@@ -430,11 +430,7 @@ plan_grouping_extension(PlannerInfo *root,
 	context.qual = qual;
 	context.numGroupCols = *p_numGroupCols;
 	context.grpColIdx = palloc0(context.numGroupCols * sizeof(AttrNumber));
-	memcpy(context.grpColIdx, *p_grpColIdx,
-		   context.numGroupCols * sizeof(AttrNumber));
 	context.grpOperators = palloc0(context.numGroupCols * sizeof(Oid));
-	memcpy(context.grpOperators, *p_grpOperators,
-		   context.numGroupCols * sizeof(Oid));
```
Comment (RekGRpth) [inline src/backend/optimizer/plan/plangroupext.c:437]: context.grpColIdx and context.grpOperators are used [below](https://github.com/arenadata/gpdb/pull/733/files#diff-7191d843b4f61084e70f78d2a438eaa2eb8989b878ec38b4ca6067097f5dea85R486)

**Comment by RekGRpth** [comment]
rebase needed

File: src/backend/optimizer/plan/plangroupext.c [hunk 2]
```diff
@@ -446,6 +442,13 @@ plan_grouping_extension(PlannerInfo *root,
 	context.canonical_rollups = NIL;
 	context.curr_grpset_no = 0;
 
+	if (*p_grpColIdx != NULL)
+		memcpy(context.grpColIdx, *p_grpColIdx,
```
Comment (RekGRpth) [inline src/backend/optimizer/plan/plangroupext.c:450]: And now context.grpColIdx and context.grpOperators can be uninitialized?

Comment (bandetto) [reply src/backend/optimizer/plan/plangroupext.c:450]: As I have checked there is one GROUP BY operator for each grouping set, so it makes sense to assert them both, as they both use `*p_numGroupCols` for their size

Comment (RekGRpth) [reply src/backend/optimizer/plan/plangroupext.c:450]: When this code is executed?
https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/optimizer/plan/plangroupext.c#L511-L519

Comment (bandetto) [reply src/backend/optimizer/plan/plangroupext.c:450]: When there is no grouping keys, eg: 
https://github.com/arenadata/gpdb/blob/cb99eab6234c784b114f4bfe0cefed03c06e379d/src/backend/cdb/cdbgroup.c#L2948-L2953

**Review by RekGRpth — APPROVED** [review]

**Pull request merged**